### PR TITLE
Airflow Standard Provider

### PR DESCRIPTION
**Pull request by romsharon98**
Extract all time operators and sensors from airflow core and remove them to `standard` provider
Mailing list thread: https://lists.apache.org/thread/2dmlqkcmyomm4q7rrovygs6bw655zx07

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: docs/apache-airflow-providers-core/index.rst
Comment (nevcohen) [inline docs/apache-airflow-providers-core/index.rst]: Required `>=2.8.0`

Comment (romsharon98) [reply docs/apache-airflow-providers-core/index.rst]: changed.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
One more thing - at least for now, "standard" should be added to list of preinstalled providers in `hatch_build.py` 

```
PRE_INSTALLED_PROVIDERS = [
    "common.compat",
    "common.io",
    "common.sql",
    "fab>=1.0.2",
    "ftp",
    "http",
    "imap",
    "smtp",
    "sqlite",
]
```

**Comment by potiuk** [comment]
We might change where it is added in Airflow 3, but for now it should be added here.

**Review by eladkal — COMMENTED** [review]
we are missing few things...
1. we need to deprecate the existed operators in core
2. we are missing all the operators and sensors we have in core (currently it's only for time operators)
3. I am not sure what about the `hooks` folder?

**Comment by potiuk** [comment]
1. we need to deprecate the existed operators in core

> Yes, but I think this should be done as separate PR to v2-10-test. I think we do not want to back-port that one to v2-10-test - and we should simply do it there (after we release `standard` as provider). I am not sure we want to back-port the whole PR here.

2. we are missing all the operators (currently it's only for time operators)

> I think those should be separate, smaller PRs ?

3. I am not sure what about the hooks folder?

> I think it's the same ^^ - some should be moved but separately.

**Comment by gopidesupavan** [comment]
Thanks @romsharon98 , will be adding rest of the operators and sensors to this.

**Comment by gopidesupavan** [comment]
> we are missing few things...
> 
> 1. we need to deprecate the existed operators in core
> 2. we are missing all the operators and sensors we have in core (currently it's only for time operators)
> 3. I am not sure what about the `hooks` folder?

@eladkal me and romsharon discussed , will be adding rest of the operators and sensors.

**Comment by potiuk** [comment]
Ah .. so no ... We should not add it to preinstalled just **yet** - until we release it (sorry @romsharon98 -> see the failure on openapi client).

**Comment by eladkal** [comment]
> Yes, but I think this should be done as separate PR to v2-10-test. I think we do not want to back-port that one to v2-10-test - and we should simply do it there (after we release standard as provider). I am not sure we want to back-port the whole PR here to v2-10-test (providers are not build nor tested there anyway).

Then I propose this PR to remove the files from main that we migrated to the provider.
The PR to v2-10-test would be just to replace the classes with deprecation notes after we release the provider.

**Comment by potiuk** [comment]
> Then I propose this PR to remove the files from main that we migrated to the provider. 

This is what the PR does.

Screenshot 2024-08-25 at 13 46 45

**Comment by eladkal** [comment]
Ah! I didn't see changes to `airflow/core` folder so I assumed it was missed.
Not very friendly UX design from GitHub

**Comment by gopidesupavan** [comment]
> > Yes, but I think this should be done as separate PR to v2-10-test. I think we do not want to back-port that one to v2-10-test - and we should simply do it there (after we release standard as provider). I am not sure we want to back-port the whole PR here to v2-10-test (providers are not build nor tested there anyway).
> 
> Then I propose this PR to remove the files from main that we migrated to the provider. The PR to v2-10-test would be just to replace the classes with deprecation notes after we release the provider.

Please correct me if i understood this, on v2-10-test deprecation updates required for all the changes we are doing on main?

**Comment by potiuk** [comment]
> Please correct me if i understood this, on v2-10-test deprecation updates required for all the changes we are doing on main?

No - only those that are supposed to affect DAG authors.

**Comment by potiuk** [comment]
BTW. We **will** have to make the old "standard" providers still work in Airflow 3 I **think** so we will have to make the old operator imports work (likely with the mechanism mentioned by @ashb - `sys.meta_path`) - otherwise it will `require` many dags to change. I think we should apply the same mechanism as we will do for importing BaseOperator etc. in Airflow 3 (and still raise deprecation warning).

@ashb ? Do you agree ?

**Comment by romsharon98** [comment]
> Interesting. The errors we see are very useful :).
> 
> 1. examples are part of the documentation of "main" airflow" - and I am afraid they will have to remain as examples in "airflow" not moved to provider, especially because point 2)...
> 2. the examples should be embedded and shown in airflow when "examples" are enabled - I think we should not (at least not now) to develop a mechanism to read examples for Airflow from providers.
> 3. The tests should also mock the moved packages.
> 
> ```
> <module 'airflow.sensors.time_delta' from '/usr/local/lib/python3.8/site-packages/airflow/sensors/time_delta.py'> does not have the attribute 'sleep'
> FAILED tests/providers/standard/time/sensors/test_time_delta.py::TestTimeDeltaSensorAsync::test_wait_sensor[True] - AttributeError: <module 'airflow.sensors.time_delta' from '/u
> ```

revert the example dags to be under `main` airflow.
should I add a link in index.yaml for those example dags? 

also not sure but should I change somewhere in the code to tell that this provider is mandatory and will auto installed?

**Comment by potiuk** [comment]
> also not sure but should I change somewhere in the code to tell that this provider is mandatory and will auto installed?

That's the preinstalled_providers - but it will only work after it's been released and it will likely change with Airflow 3 so for now - no need.

**Comment by gopidesupavan** [comment]
so for time related things its placed under time/ , for other operators and sensors am placing like this . 

standard/core/operators/
standard/core/sensors/

is this fine? or any other suggestion.

**Comment by eladkal** [comment]
Lets wait with merge till we understand the impact of https://github.com/apache/airflow/pull/41727#discussion_r1730836865

**Comment by eladkal** [comment]
> Lets wait with merge till we understand the impact of [#41727 (comment)](https://github.com/apache/airflow/pull/41727#discussion_r1730836865)

relevant code was reverted. We are good to go here.

**Review by eladkal — APPROVED** [review]

File: airflow/example_dags/example_branch_datetime_operator.py
Comment (eladkal) [inline airflow/example_dags/example_branch_datetime_operator.py:1]: you probably need to move this example to the provider

File: airflow/example_dags/example_branch_day_of_week_operator.py
Comment (eladkal) [inline airflow/example_dags/example_branch_day_of_week_operator.py:1]: you probably need to move this example to the provider

Comment (romsharon98) [reply airflow/example_dags/example_branch_day_of_week_operator.py:1]: > you probably need to move this example to the provider

as @potiuk said in the upper comment

> the examples should be embedded and shown in airflow when "examples" are enabled - I think we should not (at least not now) to develop a mechanism to read examples for Airflow from providers.

however I get a weird import error

Comment (potiuk) [reply airflow/example_dags/example_branch_day_of_week_operator.py:1]: The import error  - I explained in Slack - we migtht (likely) make the provider to be only Airflow 2.10+ compatible to allow people to change the imports in 2.10 as part of "recommended" migration process. 

I believe @ashb  planned to get the old imports to continue to work with https://docs.python.org/3/glossary.html#term-meta-path-finder with deprecation)  - I will let him comment on it.

But for the reasons explained above (examples are really part of "core" airflow and changing providers to contribute examples when "show examples" are enabled might be an overkill and since this provider is going to be "preinstalled" - I think adding a feature to read examples from providers in this case might be quite an overkill - but maybe it makes sense if you think that @eladkal -> it **could** be done possibly by just adding "providers/*/examples" to the paths when DAGs are searched. We would likely also have to figure out how to filter out "examples/system tests" in this case, because then "show examples" might led to a huge set of examples visible.

Also in this case we should move all the documentation for the standard operators to "provider" documentation.

But maybe that's a good idea?

Comment (eladkal) [reply airflow/example_dags/example_branch_day_of_week_operator.py:1]: But we do have examples for providers.. they are just not in `providers/*/examples` they are in the system tests.
We don't want `providers/*/examples` paths.. we moved all of them to the system test path.

Comment (potiuk) [reply airflow/example_dags/example_branch_day_of_week_operator.py:1]: Yeah. But currently the "examples" in core are automatically enabled when you choose to show them - and I think those "standard provider" examples should also show up - because otherwise, when we move all "core" operators to providers, there will be no more examples to show in "airflow". 

So I am ok with moving the examples to "standard provider" - but there we should figure out how to show them when we choose "show examples = True" (and we likely do not want to show all the few 100 system tests we have from other providers). Also those examples are part of the "core" documentation currently, and we should find them to be prominently visible in the docs - without having to go to "standard provider" documentation (I think) - just to make them better discoverable (currently I can imagine you know you should go to Google provider docs when you are looking for google examples - but people might not be aware they should look at standard provider when they are looking for PythonVirtualenv example - and the doc pages describing those should **likely** be in Airflow "core" docs.

So for me - this is a sign that "standard" provider's examples should be treated somewhat differently, but  I am open to hear what others think about it

Comment (eladkal) [reply airflow/example_dags/example_branch_day_of_week_operator.py:1]: I agree. I think we should favour the easiest solution to make this work. If leaving it in core is the one then so be it. We can always move them later.

Comment (gopidesupavan) [reply airflow/example_dags/example_branch_day_of_week_operator.py:1]: Agree some easiest solution would be better , and help us when we move forward down the line. 

Was refactoring bash operator to standard provider, and encountered few pre commit checks failed. i have added them to exclude block as a workoaround.  

image

image
 
Appreciate some solution or suggestion for these failures. Once the current PR merged i will raise the changes.

Comment (potiuk) [reply airflow/example_dags/example_branch_day_of_week_operator.py:1]: > Appreciate some solution or suggestion for these failures. Once the current PR merged i will raise the changes.

I think if we agree that "examples" should stay in the core, then the pre-commit should be changed to allow imports from "apache.airlfow.providers.standard" - but not for any other provider import.

**Review by eladkal — CHANGES_REQUESTED** [review]
Blocking merge till we sort out what we want this provider to be

https://github.com/apache/airflow/pull/41736#issuecomment-2313215620

File: airflow/example_dags/example_branch_day_of_week_operator.py
Comment (gopidesupavan) [reply airflow/example_dags/example_branch_day_of_week_operator.py:1]: Thanks Jarek :) , makes sense.

**Comment by romsharon98** [comment]
I think that after the [consensus](https://lists.apache.org/thread/1ltpczo121tgk2pzx31bgsk0zj4s5o6x) we can remove the request change @eladkal.

**Review by potiuk — APPROVED** [review]

File: airflow/providers/standard/CHANGELOG.rst
Comment (potiuk) [inline airflow/providers/standard/CHANGELOG.rst]: We should make sure to add a note for every deprecation we added vs. the original providers. Basically this changeliog shouid describe what has changed here comparing to the operators/sensors/hooks we had in `Airflow 2`. 

The format of this should be like "this and this deprecated thing has been removed, you need to do this and that to fix it" (i.e. actionable decription of the change).

File: dev/breeze/src/airflow_breeze/global_constants.py
Comment (potiuk) [inline dev/breeze/src/airflow_breeze/global_constants.py]: I think we should aim to get standard to be **not** removed. Not sure why we are removing it now - but we agreed that 2.8 should be the "min-version" for standard provider, so the tests should be adapted to work for those versions as well..

File: generated/provider_dependencies.json
Comment (potiuk) [inline generated/provider_dependencies.json:1266]:
```suggestion
      "apache-airflow>=2.8.0"
```

File: airflow/providers/standard/provider.yaml
Comment (potiuk) [inline airflow/providers/standard/provider.yaml:31]:
```suggestion
  - apache-airflow>=2.8.0
```

**Comment by potiuk** [comment]
LGTM in general.  Tests are failing and need to be fixed, and I added few comments that should be resolved before merging that one. Also I think we should figure out (but that might be next) how to get example_dags to display them in Airflow 3 from provider (likely change/extend from where they are loaded) - but this should be a separate PR, likely when we have few more examples moved.

**Comment by gopidesupavan** [comment]
@romsharon98 just checking is the changes required in here https://github.com/apache/airflow/blob/main/docs/apache-airflow/operators-and-hooks-ref.rst?plain=1#L91 ?

**Comment by romsharon98** [comment]
> @romsharon98 just checking is the changes required in here https://github.com/apache/airflow/blob/main/docs/apache-airflow/operators-and-hooks-ref.rst?plain=1#L91 ?

good catch! I will modify it 😄

File: dev/breeze/src/airflow_breeze/global_constants.py
Comment (romsharon98) [reply dev/breeze/src/airflow_breeze/global_constants.py]: I remember why I did it and kept the min version to be 2.10.
We talked about it in slack: https://apache-airflow.slack.com/archives/C06K9Q5G2UA/p1724737809230079

**Review by eladkal — APPROVED** [review]

**Pull request merged**
